### PR TITLE
[3.13] gh-122695: Fix double-free when using `gc.get_referents` with a freed `_asyncio.FutureIter`

### PR DESCRIPTION
**Pull request by ZeroIntensity**
CC @colesbury

* Issue: gh-122695

### CODE REVIEW

**Comment by colesbury** [comment]
I think it's worth including the test case from #122834

**Comment by ZeroIntensity** [comment]
> I think it's worth including the test case from #122834

Oh yeah, forgot to add that here as well. It's there now.

File: Lib/test/test_asyncio/test_futures.py [hunk 1]
```diff
@@ -675,6 +675,13 @@ def test_future_del_segfault(self):
         with self.assertRaises(AttributeError):
             del fut._log_traceback
 
+    def test_future_iter_get_referents_segfault(self):
+        # See https://github.com/python/cpython/issues/122695
+        import _asyncio
+        it = iter(self._new_future(loop=self.loop))
+        del it
+        evil = gc.get_referents(_asyncio)
```
Comment (vstinner) [inline Lib/test/test_asyncio/test_futures.py:683]: You should force a `gc.collect()` here.

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -3530,16 +3530,8 @@ module_traverse(PyObject *mod, visitproc visit, void *arg)
...
-    PyObject *next = (PyObject*) state->fi_freelist;
-    while (next != NULL) {
-        PyObject *current = next;
-        Py_VISIT(current);
-        next = (PyObject*) ((futureiterobject*) current)->future;
-    }
-#endif
-
+    // GH-122695: Do not traverse the freelist here, as that can cause problems
+    // with gc.get_referents()
```
Comment (vstinner) [inline Modules/_asynciomodule.c]: I don't think that this comment is useful, I suggest to just remove it.

File: Lib/test/test_asyncio/test_futures.py [hunk 3]
```diff
@@ -675,6 +675,14 @@ def test_future_del_segfault(self):
         with self.assertRaises(AttributeError):
             del fut._log_traceback
 
+    def test_future_iter_get_referents_segfault(self):
+        # See https://github.com/python/cpython/issues/122695
+        import _asyncio
+        it = iter(self._new_future(loop=self.loop))
+        del it
+        gc.collect()
```
Comment (vstinner) [inline Lib/test/test_asyncio/test_futures.py]: The collect() must be done once evil list is created, not before.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Changes in the 3.13 branch currently need approvals from two core devs.

**Review by kumaraditya303 — APPROVED** [review]

**Review by colesbury — APPROVED** [review]

**Pull request merged**